src/Temporalio/Bridge/src/runtime.rs

```diff
@@ -0,0 +1,223 @@
+use std::sync::Arc;
+use std::time::Duration;
+use temporal_sdk_core::CoreRuntime;
+use temporal_sdk_core_api::telemetry::{
+    Logger, MetricsExporter, OtelCollectorOptions, TelemetryOptions as CoreTelemetryOptions,
+    TelemetryOptionsBuilder, TraceExportConfig, TraceExporter,
+};
+use url::Url;
+
+#[repr(C)]
+pub struct RuntimeOptions {
+    telemetry: *const TelemetryOptions,
+}
+
+#[repr(C)]
+pub struct TelemetryOptions {
+    tracing: *const TracingOptions,
+    logging: *const LoggingOptions,
+    metrics: *const MetricsOptions,
+}
+
+#[repr(C)]
+pub struct TracingOptions {
+    filter: ByteArrayRef,
+    opentelemetry: OpenTelemetryOptions,
+}
+
+#[repr(C)]
+pub struct LoggingOptions {
+    filter: ByteArrayRef,
+    forward: bool,
+}
+
+#[repr(C)]
+pub struct MetricsOptions {
+    opentelemetry: *const OpenTelemetryOptions,
+    prometheus: *const PrometheusOptions,
+}
+
+#[repr(C)]
+pub struct OpenTelemetryOptions {
+    url: ByteArrayRef,
+    /// Headers are <key1>\n<value1>\n<key2>\n<value2>. Header keys or values
+    /// cannot contain a newline within.
+    headers: ByteArrayRef,
+    metric_periodicity_millis: u32,
+}
+
+#[repr(C)]
+pub struct PrometheusOptions {
+    bind_address: ByteArrayRef,
+}
+
+#[derive(Clone)]
+pub struct Runtime {
+    pub(crate) core: Arc<CoreRuntime>,
+}
+
+/// If success or fail are not null, they must be manually freed when done.
+/// Runtime is always present, but it should never be used if fail is present,
+/// only freed after fail is freed.
+#[repr(C)]
+pub struct RuntimeOrFail {
+    runtime: *mut Runtime,
+    fail: *const ByteArray,
+}
+
+#[no_mangle]
+pub extern "C" fn runtime_new(options: *const RuntimeOptions) -> RuntimeOrFail {
+    match Runtime::new(unsafe { &*options }) {
+        Ok(runtime) => RuntimeOrFail {
+            runtime: Box::into_raw(Box::new(runtime)),
+            fail: std::ptr::null(),
+        },
+        Err(err) => {
+            // We have to make an empty runtime just for the failure to be
+            // freeable
+            let mut runtime = Runtime {
+                core: Arc::new(
+                    CoreRuntime::new(
+                        CoreTelemetryOptions::default(),
+                        tokio::runtime::Builder::new_current_thread(),
+                    )
+                    .unwrap(),
+                ),
+            };
+            let fail = runtime.alloc_utf8(&format!("Invalid options: {}", err));
+            RuntimeOrFail {
+                runtime: Box::into_raw(Box::new(runtime)),
+                fail: fail.into_raw(),
+            }
+        }
+    }
+}
+
+#[no_mangle]
+pub extern "C" fn runtime_free(runtime: *mut Runtime) {
+    unsafe {
+        let _ = Box::from_raw(runtime);
+    }
+}
+
+#[no_mangle]
+pub extern "C" fn byte_array_free(runtime: *mut Runtime, bytes: *const ByteArray) {
+    // Bail if freeing is disabled
+    unsafe {
+        if bytes.is_null() || (*bytes).disable_free {
+            return;
+        }
+    }
```

Comment:
Oh. Here it is, yeah, would be good to reference this from `ByteArray`'s `Drop`

Reply:
As in via comment? I can, sure. But yeah I don't want this logic inside of drop because "disable free" is a construct of the runtime use of this not the actual data. I still want drop to drop everything it has when it is dropped.

Reply:
Yes, just to make it clear. I understand why it needs to happen outside.

Reply:
Fixed